File: lab7/schedule.hs
import Data.List
import Data.Foldable
import Data.Function
import System.CPUTime
import Text.Printf

data Task = Task { taskId :: Int, taskTime :: Int }
data Machine = Machine { totalTime :: Int, assignedTasks :: [Int] }

consWhen :: Bool -> a -> [a] -> [a]
consWhen p a = if p then (a :) else id

newMachine :: Task -> Machine
newMachine (Task i t) = Machine t [i]

updateMachine :: Task -> Machine -> Machine
updateMachine (Task i t) (Machine t' i') = Machine (t + t') (i : i')


assignTask :: Int -> Int -> Task -> [Machine] -> [[Machine]]
assignTask k limit task machines =
  consWhen (length machines < k) (newMachine task : machines) (f machines)
  where
    f [] = []
    f (m : ms) = consWhen (totalTime m + taskTime task <= limit)
                    (updateMachine task m : ms) [m : l | l <- f ms]


strategies :: [Task] -> Int -> Int -> [[Machine]]
strategies tasks k limit = f tasks []
  where
    f [] machines = [machines]
    f (t : ts) machines = concatMap (f ts) (assignTask k limit t machines)


searchMinTime :: [Task] -> Int -> Int -> Int -> [[Machine]]
searchMinTime tasks k l r =
  case compare l r of
    GT -> undefined
    EQ -> strategies tasks k r
    LT -> let mid = (l + r) `div` 2 in
          if null (strategies tasks k mid)
            then searchMinTime tasks k (mid + 1) r
            else searchMinTime tasks k l mid


bestStrategies :: [Task] -> Int -> [[Machine]]
bestStrategies tasks k = searchMinTime sorted k l r
  where
    sorted = sortBy (flip compare `on` taskTime) tasks
    times = map taskTime tasks
    l = sum times `div` k
    r = sum times


runTest :: Int -> IO ()
runTest c = do
  printf "Running test %d\n" c
  _ : k : times <- map read . words <$> readFile (printf "test%d.txt" c)
  let tasks = zipWith Task [1..] times
  let machines = head $ bestStrategies tasks k
  start <- getCPUTime
  machines `seq` return ()
  end <- getCPUTime
  let diff = fromIntegral (end - start) / (10^12)
  printf "Algorithm completed in %.6fs\n" (diff :: Double)
  printf "Solution:\n"
  printf "Min time required: %d\n" (maximum $ map totalTime machines)
  printf "A strategy:\n"
  forM_ (zip [1..] machines) $ \(i, Machine _ t) -> do
    printf "  Machine %d: tasks = %s\n" (i :: Int) (show $ sort t)
  printf "\n"


main :: IO ()
main = mapM_ runTest [1, 2, 3]

File: lab7/schedule.c
#include <stdio.h>
#include <stdbool.h>
#include <string.h>
#include <time.h>

#define TASKS_MAX 30
#define MACHINES_MAX 20

int n, k;
int task[TASKS_MAX];

int min;
unsigned long assignedTasks[MACHINES_MAX];

static int id[TASKS_MAX];
static int total[MACHINES_MAX];
static int curr;

static void findMinTime(int i, int j)
{
    if (i == n) {
        min = curr;
        return;
    }

    if (j < k) {
        total[j] = task[i];
        findMinTime(i + 1, j + 1);
    }

    for (int l = 0; l < j; ++l) {
        if (total[l] + task[i] < min) {
            total[l] += task[i];
            int save = curr;
            if (total[l] > curr)
                curr = total[l];

			findMinTime(i + 1, j);

            total[l] -= task[i];
            curr = save;

            if (curr == min)
                break;
        }
    }
}

static bool findStrategy(int i, int j)
{
    if (i == n)
        return true;

    if (j < k) {
        total[j] = task[i];

        if (findStrategy(i + 1, j + 1)) {
            assignedTasks[j] |= 1 << id[i];
            return true;
        }
    }

    for (int l = 0; l < j; ++l) {
        if (total[l] + task[i] <= min) {
            total[l] += task[i];

            if (findStrategy(i + 1, j)) {
                assignedTasks[l] |= 1 << id[i];
                return true;
            }

            total[l] -= task[i];
        }
    }

    return false;
}

inline static void sortTask(void)
{
    for (int i = 0; i < n; ++i)
        id[i] = i;

    for (int i = 1; i < n; ++i) {
        int t = task[i], o = id[i];

        int j = i;
        for (; task[j - 1] < t && j > 0; --j) {
            task[j] = task[j - 1];
            id[j] = id[j - 1];
        }

        task[j] = t;
        id[j] = o;
    }
}

void schedule(void)
{
    sortTask();

    min = 0;
    for (int i = 0; i < n; i += k)
        min += task[i];

    total[0] = task[0];
    curr = task[0];
    findMinTime(1, 1);

    memset(assignedTasks, 0, sizeof(assignedTasks));
    total[0] = task[0];
    assignedTasks[0] = 1 << id[0];
    findStrategy(1, 1);
}

void test(int num)
{
    printf("Running test %d\n", num);

    char filename[20];
    snprintf(filename, sizeof(filename), "test%d.txt", num);
    FILE *fp = fopen(filename, "r");

    fscanf(fp, "%d%d", &n, &k);
    for (int i = 0; i < n; ++i) {
        fscanf(fp, "%d", &task[i]);
    }

    clock_t start = clock();
    schedule();
    clock_t end = clock();

    double diff = (double) (end - start) / CLOCKS_PER_SEC;
    printf("Algorithm completed in %.6lf seconds, solution:\n", diff);

    printf("Min time required: %d\n", min);
    printf("A strategy:\n");
    for (int i = 0; i < k; ++i) {
        printf("  Machine %d: tasks = {", i + 1);
        for (int j = 0; j < n; ++j) {
            if (assignedTasks[i] & (1 << j)) {
                printf("%d, ", j + 1);
            }
        }
        printf("}\n");
    }
    printf("\n");

    fclose(fp);
}

int main(void)
{
    test(1);
    test(2);
    test(3);
    return 0;
}
